### PR TITLE
Remove nested loop in pruneInitContainersBeforeStart of kuberuntime_container

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Currently we use nested loop in pruneInitContainersBeforeStart to prune InitContainer. This unnecessarily loops podStatus.ContainerStatuses where there is no single match in initContainerNames Set.
This PR removes the outer loop. We can use a map to record the number of times any initContainer name has been seen and prune accordingly.

Same change is applied to purgeInitContainers as well.

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 1]
```diff
@@ -626,39 +625,40 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, ru
...
+			continue
+		}
+		if (status.State != kubecontainer.ContainerStateExited &&
+				status.State != kubecontainer.ContainerStateUnknown) {
+			continue
+		}
+		// Remove init containers in unknown state. It should have
+		// been stopped before pruneInitContainersBeforeStart is
+		// called.
+		count := initContainerNames[status.Name]
```
Comment (mattjmcnaughton) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]:
```suggestion
`initContainerNames[status.Name] += 1
```
Can we replace these three lines with the above suggestion?

**Review by mattjmcnaughton — APPROVED** [review]
Thanks for this pr :)

Please fix the `gofmt` error, but otherwise, just one small suggested change that you can make if you want.

/lgtm

**Comment by tedyu** [comment]
@mattjmcnaughton 
Can you give lgtm again ?

**Comment by mattjmcnaughton** [comment]
/lgtm

nice work :)

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by tedyu** [comment]
@yujuhong
Can you take a look ?

**Comment by tedyu** [comment]
ping @yujuhong for review

**Comment by tedyu** [comment]
cc @dchen1107

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by tedyu** [comment]
@yujuhong
Can you take a look ?

**Comment by tedyu** [comment]
@vishh 
Please review.

**Comment by tedyu** [comment]
@vishh
Please take a look.

**Comment by tedyu** [comment]
@vishh
Please take a look.

**Comment by tedyu** [comment]
ping @vishh for review.

**Comment by tedyu** [comment]
ping @derekwaynecarr for review

**Comment by tedyu** [comment]
@vishh 
Please review

**Comment by tedyu** [comment]
@vishh
Please review

**Comment by tedyu** [comment]
@vishh
Please review

**Comment by tedyu** [comment]
@vishh
Please review

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by mattjmcnaughton** [comment]
@tedyu I think it could be helpful to to
/assign @vishh 
instead of pinging them.

In addition, if a certain reviewer isn't responding, sometimes its helpful to choose another from `pkg/kubelet/OWNERS`.

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by tedyu** [comment]
ping @vishh for review

**Comment by tedyu** [comment]
In favor of #83245

**Pull request closed without merging**